federatedscope/cl/trainer/trainer.py

```diff
@@ -0,0 +1,144 @@
+        for data, target in test_bar:
+            data, target = data.to(device), target.to(device)
+            feature = net(data)
+            feature = F.normalize(feature, dim=1)
+            
+            pred_labels = knn_predict(feature, feature_bank, feature_labels, classes, k, t)
+
+            total_num += data.size(0)
+            total_top1 += (pred_labels[:, 0] == target).float().sum().item()
+    return total_top1 / total_num * 100
+
+
+class CLTrainer(GeneralTorchTrainer):
+    def _hook_on_batch_forward(self, ctx):
+        x, label = [utils.move_to(_, ctx.device) for _ in ctx.data_batch]
+#         print(len(x), x[0].size(), x[1].size(), label.size())
+        x1, x2 = x[0], x[1]
+        z1, z2 = ctx.model(x1, x2)
+        if len(label.size()) == 0:
+            label = label.unsqueeze(0)
+        ctx.loss_batch = ctx.criterion(z1, z2)
+        ctx.y_true = label
+        ctx.y_prob = z1, z2
+
+        ctx.batch_size = len(label)
+        
+    def _hook_on_batch_end(self, ctx):
+        # update statistics
+        setattr(
+            ctx, "loss_batch_total_{}".format(ctx.cur_data_split),
+            ctx.get("loss_batch_total_{}".format(ctx.cur_data_split)) +
+            ctx.loss_batch.item() * ctx.batch_size)
+
+        if ctx.get("loss_regular", None) is None or ctx.loss_regular == 0:
+            loss_regular = 0.
+        else:
+            loss_regular = ctx.loss_regular.item()
+        setattr(
+            ctx, "loss_regular_total_{}".format(ctx.cur_data_split),
+            ctx.get("loss_regular_total_{}".format(ctx.cur_data_split)) +
+            loss_regular)
+        setattr(
+            ctx, "num_samples_{}".format(ctx.cur_data_split),
+            ctx.get("num_samples_{}".format(ctx.cur_data_split)) +
+            ctx.batch_size)
+
+        # cache label for evaluate
+        ctx.get("{}_y_true".format(ctx.cur_data_split)).append(
+            ctx.y_true.detach().cpu().numpy())
+
+#         print(len(ctx.y_prob), ctx.y_prob[0].size(), ctx.y_prob[1].size())
+        ctx.get("{}_y_prob".format(ctx.cur_data_split)).append(
+            ctx.y_prob[0].detach().cpu().numpy())
+
+        # clean temp ctx
+        ctx.data_batch = None
+        ctx.batch_size = None
+        ctx.loss_task = None
+        ctx.loss_batch = None
+        ctx.loss_regular = None
+        ctx.y_true = None
+        ctx.y_prob = None
+        
+    def _hook_on_fit_end(self, ctx):
+        """Evaluate metrics.
+
+        """
+        setattr(
+            ctx, "{}_y_true".format(ctx.cur_data_split),
+            np.concatenate(ctx.get("{}_y_true".format(ctx.cur_data_split))))
+        setattr(
+            ctx, "{}_y_prob".format(ctx.cur_data_split),
+            np.concatenate(ctx.get("{}_y_prob".format(ctx.cur_data_split))))
+        results = self.metric_calculator.eval(ctx)
+        setattr(ctx, 'eval_metrics', results)
+
+def linear_prob_forward(ctx):
+    x, label = [_.to(ctx.device) for _ in ctx.data_batch]
```

Comment:
why do we need to traverse a mini-batch instead of processing in a batching mode